kythe/go/util/riegeli/reader.go

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package riegeli
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+)
+
+func (r *Reader) nextRecord() ([]byte, error) {
+	if err := r.readNextRecordChunk(); err != nil {
+		return nil, err
+	}
+
+	size, err := binary.ReadUvarint(r.sizesReader)
+	if err != nil {
+		return nil, fmt.Errorf("error reading record size: %v", err)
+	}
+	val := make([]byte, size)
+	if _, err := io.ReadFull(r.valsReader, val); err != nil {
+		return nil, fmt.Errorf("error reading record value: %v", err)
+	}
+	r.numRecords--
+	return val, nil
+}
+
+func (r *Reader) readNextRecordChunk() error {
+	for r.valsReader == nil || r.numRecords == 0 {
+		c, err := r.r.Next()
+		if err != nil {
+			return err
+		} else if c.Header.ChunkType != recordChunkType {
+			// TODO(schroederc): support non-record chunk types (especially RecordsMetadata)
```

Comment:
Do these occur often enough to be worth logging?

Reply:
Maybe once I implement the RecordsMetadata and transposed chunks.  For now, it should happen about once per file for the file signature.